packages/arb-rpc-node/blockcache/blockcache.go

```diff
@@ -0,0 +1,137 @@
+package blockcache
+
+import (
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/offchainlabs/arbitrum/packages/arb-rpc-node/snapshot"
+	"sync"
+	"time"
+)
+
+type record struct {
+	Header   *types.Header
+	Snapshot *snapshot.Snapshot
+}
+
+type BlockCache struct {
+	expiration time.Duration
+
+	lock         sync.RWMutex
+	cache        map[uint64]*record
+	nextHeight   uint64
+	oldestHeight uint64
+}
+
+func New(initialSize int, expiration time.Duration) (*BlockCache, error) {
+	return &BlockCache{
+		expiration: expiration,
+		cache:      make(map[uint64]*record, initialSize),
+	}, nil
+}
+
+func (bc *BlockCache) Size() int {
+	return len(bc.cache)
+}
+
+// emptyCacheNoLock removes all entries
+func (bc *BlockCache) emptyCacheNoLock() {
+	bc.cache = make(map[uint64]*record, len(bc.cache))
+	bc.oldestHeight = 0
+	bc.nextHeight = 0
+}
+
+// reorgNoLock removes all obsolete blocks up to and including nextHeight
+func (bc *BlockCache) reorgNoLock(nextHeight uint64) {
+	if bc.nextHeight == 0 {
+		// Nothing to remove
+		return
+	}
+
+	if nextHeight <= bc.oldestHeight {
+		// Remove everything
+		bc.emptyCacheNoLock()
+		return
+	}
+
+	for currentHeight := bc.nextHeight - 1; currentHeight >= nextHeight; currentHeight-- {
+		if _, exists := bc.cache[currentHeight]; !exists {
+			// No cache entry for currentHeight
+			continue
```

Comment:
This check isn't necessary as `delete` is a noop if the key doesn't exist

Reply:
:+1: